odl/trafos/fourier.py

```diff
@@ -0,0 +1,1024 @@
+                             'with complex input.')
+        else:
+            out_dtype = arr_in.dtype
+
+        if arr_out.dtype != out_dtype:
+            raise TypeError('Expected output dtype {}, got {}.'
+                            ''.format(out_dtype, arr_out.dtype))
+
+    elif direction == 'backward':
+        in_shape = list(arr_out.shape)
+        if halfcomplex:
+            in_shape[axes[-1]] = arr_out.shape[axes[-1]] // 2 + 1
+
+        if arr_in.shape != tuple(in_shape):
+            raise ValueError('Expected input shape {}, got {}.'
+                             ''.format(tuple(in_shape), arr_in.shape))
+
+        if is_real_dtype(arr_out.dtype):
+            in_dtype = _TYPE_MAP_R2C[arr_out.dtype]
+        elif halfcomplex:
+            raise ValueError('Cannot combine halfcomplex backward transform '
+                             'with complex output.')
+        else:
+            in_dtype = arr_out.dtype
+
+        if arr_in.dtype != in_dtype:
+            raise TypeError('Expected input dtype {}, got {}.'
+                            ''.format(in_dtype, arr_in.dtype))
+
+    else:  # Shouldn't happen
+        raise RuntimeError
+
+
+def pyfftw_call(array_in, array_out, direction='forward', axes=None,
+                halfcomplex=False, **kwargs):
+    """Calculate the DFT.
+
+    The discrete Fourier transform calcuates the sum
+
+        :math:`\widehat{f}_k =
+        \sum_{j \\in I_N} f_j\ e^{-i 2\pi j\odot k / N}`
+
+    for indices :math:`k \\in I_N` or, in the half-complex case,
+    :math:`0 \\leq k_d \\leq \\lfloor N_d / 2 \\rfloor + 1` for the
+    last component.
+
+    Parameters
+    ----------
+    array_in : `numpy.ndarray`
+        Array to be transformed
+    array_out : `numpy.ndarray`
+        Output array storing the transformed values
+    direction : {'forward', 'backward'}
+        Direction of the transform
+    axes : sequence of `int`, optional
+        Dimensions along which to take the transform. `None` means
+        using all axis and is equivalent to ``np.arange(ndim)``.
+    halfcomplex : `bool`, optional
+        If `True`, calculate only the negative frequency part along the
+        last axis. If `False`, calculate the full complex FFT.
+        This option can only be used with real input data.
+    fftw_plan : ``pyfftw.FFTW``, optional
+        Use this plan instead of calculating a new one. If specified,
+        the options ``planning_effort``, ``planning_timelimit`` and
+        ``threads`` have no effect.
+    planning_effort : {'estimate', 'measure', 'patient', 'exhaustive'}
+        Flag for the amount of effort put into finding an optimal
+        FFTW plan. See the `FFTW doc on planner flags
+        <http://www.fftw.org/fftw3_doc/Planner-Flags.html>`_.
+    planning_timelimit : `float`, optional
+        Limit planning time to roughly this amount of seconds.
+        Default: `None` (no limit)
+    threads : `int`, optional
+        Number of threads to use. Default: 1
```

Comment:
I'm not aware of an option to simply say "use all". You need to give the number explicitly. And on my machine, 3 threads were faster than 4 but I have 8 cores?? Perhaps there's some clever way to find out the number of cores and use a reasonable number based on that.